### ISSUE TITLE
Disconnection after 1 minute

### ISSUE THREAD

**Issue by piranna** [because-why-not/awrtc_signaling#8]
With my client from #6, everything is working ok now and can be able to get the WebRTC streams, but Unity clients send a `Disconnect` event about one minute after connecting. Seems like there's some kind of timeout or heartbeat that I'm not considering on my Node.js client, but I've been printing all received messages and nothing is being shown, and heartbeat are correctly processed and answered but don't receive any of them. Is there something I'm missing in the protocol? Why Unity clients could be disconnecting?

**Comment by devluz** [comment]
Any log messages / signaling messages you could show me? 

The 60 second timeout is triggered if after the time no full connection is established. For example the asset might still be waiting for the data channels to open.

**Comment by piranna** [comment]
> Any log messages / signaling messages you could show me?

Not from Unity, it shows nothing. Server side everything looks ok.


> The 60 second timeout is triggered if after the time no full connection is established. For example the asset might still be waiting for the data channels to open.

How can this be? Video is being send correctly, is seems everything is working ok, and this looks like an additional restriction over regular WebRTC. Why is that? How can be check what's missing? How can be disabled?

**Comment by devluz** [comment]
If you use the Unity CallApp example it has verbose mode on by default and should print all messages it receives and sends. Alternatively, the signaling server logs them as well. 

Yes it has many additional restrictions over WebRTC. The asset was originally developed in 2015 and since then there were many different WebRTC implementation & versions that were partially incompatible to each other. Some working only on specific browsers or platforms.
The asset focuses on a very specific configuration for WebRTC to make sure all features of the ICall API work on all the different platforms & implementation it was designed for and they are all acting the exact same way to keep them compatible to each other. The native platforms for Unity C# have no access to a proper WebRTC API like in the browser so if the sender side triggers WebRTC events that don't exist on the other side you run into problems like these. 

I might be able to see the problem if you show me the signaling messages your are exchanging with the asset. A test case for me to reproduce the problem would be better though.

**Comment by piranna** [comment]
Hi @devluz, sorry for the delay, I was waiting for aproval from my boss to share my code with you.

At https://github.com/because-why-not/awrtc_signaling/pull/9 I've done a pull-request with the code of the Node.js client I've been developing during the last weeks, and at https://github.com/piranna/awrtc_node-webrtc I have the proof of concept I've been working on that makes use of it (it's just under 200 sloc, so it will be easy to follow and understand). With this last code I've been able to get the screen content using Unity CallApp and stream it using HLS. Since you talk about `awrtc` module is focused on that particular use case, maybe the problem could be related to the fact that `awrtc_node-webrtc` project is just only receiving a stream and not emitting it, so if CallApp maybe that's the reason of the timeout, since we are not sending anything to it. Can you confirm this point, and if so, suggest us a solution? Maybe Unity plugin can be changed so this timeout check is optional?

**Comment by devluz** [comment]
My first guess would be the following:
The asset uses the old AddTrack to be compatible to (very old) native clients still using the AddStream method: https://github.com/because-why-not/awrtc_browser/blob/master/src/awrtc/media_browser/MediaPeer.ts#L131
It gets the track & a stream reference

This is probably incompatible to addTransceiver:
https://github.com/piranna/awrtc_node-webrtc/blob/main/index.js#L158 which lacks the connection between stream & track. It is possible you can add this on your side though via RTCRtpTransceiverInit.  I can probably also fix it on my side by allowing tracks without stream. 

I will have a closer look into it this weekend.
Just so I understand correctly how it is used: The node.js client will stream a test video on address and the other clients should be able to receive it by just connecting to that address. Is that correct?

**Comment by piranna** [comment]
> The node.js client will stream a test video on address and the other clients should be able to receive it by just connecting to that address. Is that correct?

No, just the other way: Unity clients are used unmodified, connecting to the `awrtc_signaling` server and they are streaming to each others as usual.

The Node.js client connect to the `awrtc_signaling` server too, BUT it's not emitting any stream, just only listening and receiving the streams from the Unity clients. Receiving request is initiated by the Node.js client, and that's the reason why I'm using `addTransceiver()`, to create a `recv-only` SDP `offer`, since the old non-standard `offerToReceiveAudio` and `offerToReceiveVideo` has been long time deprecated and removed. Returned `answer` SDP is `send-only`, so it seems the Unity clients are identifying them correctly and shouldn't be waiting for anything else... I don't think it's something streams related since I'm already receiving them, but instead some unneeded checks done inside the `awrtc` Unity plugin.

Please keep me updated on your progress on this topic during this weekend, and tell me if we can help you someway with it.

**Comment by devluz** [comment]
No luck running the awrtc_node-webrtc so far. Does it need any specific nodejs version or OS to run on? 
npm install returns:

> npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
> npm ERR! code ENOLOCAL
> npm ERR! Could not install from "node_modules\unify-config\caller-callsite@github:piranna\caller-callsite" as it does not contain a package.json file.

Removing the package-lock.json works but npm start gives me the following:

> class ClientEvent extends Event {
>                           ^
> 
> ReferenceError: Event is not defined
>     at Object.<anonymous> (C:\dev\wrtc\bugs\2021-02-02_github_typescript_incmpatibility_nodejs\awrtc_node-webrtc\node_modules\awrtc_signaling\client.js:40:27)
>     at Module._compile (internal/modules/cjs/loader.js:1063:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
>     at Module.load (internal/modules/cjs/loader.js:928:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:769:14)
>     at ModuleWrap.<anonymous> (internal/modules/esm/translators.js:199:29)
>     at ModuleJob.run (internal/modules/esm/module_job.js:152:23)
>     at async Loader.import (internal/modules/esm/loader.js:166:24)
>     at async Object.loadESM (internal/process/esm_loader.js:68:5)
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1

**Comment by piranna** [comment]
Yes, you need Node.js 15, that includes both npm 7 (with `package-lock.json` version 2 format) and public `Event` and `EventTarget` APIs.

**Comment by piranna** [comment]
By the way, I'm developing on Ubuntu 20.10 on x86_64, not Windows, but I don't think this would affect you.

**Comment by devluz** [comment]
Ok it works on Ubuntu 20. It didn't work on windows at all though. 

The main issue is that the asset expects the reliable & unreliable data channel to open before it reports back an opened connection.  If this doesn't happen it runs into the timeout waiting for it. It is done this way to make sure the ICall.Send methods actually work when the user receives the connection event. 

So easiest fix might be to do this on your side:
https://github.com/because-why-not/awrtc_browser/blob/master/src/awrtc/network/WebRtcPeer.ts#L557

I am also testing to connect without data channels in the V0.985 builds via a NativeMediaConfig.UseDataChannels flag. I am not yet sure if I really want to add though.

**Comment by piranna** [comment]
> Ok it works on Ubuntu 20. It didn't work on windows at all though.

Maybe due to `node-webrtc`?


> So easiest fix might be to do this on your side:

Create two DataChannels, althought I would not make use of them?


> I am also testing to connect without data channels in the V0.985 builds via a NativeMediaConfig.UseDataChannels flag. I am not yet sure if I really want to add though.

I find it better that these features are opt-in instead of opt-out, in fact Kurento added DataChannels support this way, by adding a flag to enable this functionality. In any case, the optimas solution would be to don't need a flag at all and both detect the DataChannel creation request and/or don't wait for preconditions at all and just notify them similar to how browsers does, but maybe that would means a change in the API.

Just by curiosity, what libwebrtc version are you using? What Chrome version is it equivalent to?

**Comment by devluz** [comment]
> Create two DataChannels, althought I would not make use of them?

Yes I am afraid this means creating two data channels even if they aren't used. WebRTC Video Chat is already 6 years old. It started as WebRTC Network (mostly a text chat) and added video & audio features over time. There are several parallel implementations in different languages so to keep them all compatible the exact peer configuration is kept fixed. 

> I find it better that these features are opt-in instead of opt-out, in fact Kurento added DataChannels support this way, by adding a flag to enable this functionality.

I am afraid it is too late for that. I try to keep the asset compatible so users applications do not break when they update. I think for now the UseDataChannels flag should work well enough. Soon we will have a proper C# WebRTC API in Unity and then it doesn't matter anymore anyway. Until then the asset needs to remain as simple as possible to keep it maintainable for so many different platforms with just one developer (and zero external funding)

> Just by curiosity, what libwebrtc version are you using? What Chrome version is it equivalent to?

The current asset store version is on 72 (following Microsofts WebRTC UWP implementation) and the update that is coming now will be on m87 and should be following chromium closely now that native API's aren't breaking every month anymore. 

Did you get it running on your side now? Test builds for V0.985 are available here: https://www.because-why-not.com/files/dl/?f=f&dir=build but if you want to be compatible to older existing apps it is best to stick to two data channels.

**Comment by piranna** [comment]
> I am afraid it is too late for that. I try to keep the asset compatible so users applications do not break when they update.

Use [semver](https://semver.org/), maybe? That would easily notify and prevent users that there would be breaking changes.


> Did you get it running on your side now? Test builds for V0.985 are available here: [because-why-not.com/files/dl/?f=f&dir=build](https://www.because-why-not.com/files/dl/?f=f&dir=build) but if you want to be compatible to older existing apps it is best to stick to two data channels.

I didn't try it myself, I'll ask my colleagues to update their dependency and give it a try. Compatibility with older apps is not an issue because it's a new project, but maybe I would need to do that if C# team doesn't have time to do the upgrade.

**Comment by piranna** [comment]
We have downloaded latest version of the plugin but it's not crear how we can set the `UseDataChannel` flag, nor find it documented. Does we need to set it here? How? Can you provide an example?

```csharp
protected virtual ICall CreateCall(NetworkConfig netConfig){
    //setup the server
    return UnityCallFactory.Instance.Create(netConfig);
}
```

**Comment by devluz** [comment]
Sorry forgot to mention. It is currently in NativeMediaConfig. Watch out the last builds are still unstable. It could still take a few weeks until it gets into the asset store. Alone the approval process can take a week so this isn't something that can be quickly updated or changed like a webpage or npm package. 

```csharp
    NativeMediaConfig conf = new NativeMediaConfig();
    conf.UseDataChannels = false;
    //[...]
    mCall.Configure(conf );
```
I can also send out examples but best contact me via e-mail about that.

**Comment by piranna** [comment]
> Sorry forgot to mention. It is currently in NativeMediaConfig.

No worries :-) We have already tested that but it's not working. We have updated the Unity plugin and the flag is recognized when compiling the C# client, but we are still getting the disconnection, not sure why. In fact, with new version I'm only getting the audio track, not the video one... :-/


> I can also send out examples but best contact me via e-mail about that.

My boss Francesco D'Andria asked me for your email details some days ago, not sure if he has already contact you yet, if not I can ask him to contact for the details.

**Comment by devluz** [comment]
I don't think I have received an e-mail yet. Best way to reach me is via contact@because-why-not.com . I am about to finish the next release. After that there is more time for any new changes.

**Comment by piranna** [comment]
> I don't think I have received an e-mail yet. Best way to reach me is via [contact@because-why-not.com](mailto:contact@because-why-not.com) . I am about to finish the next release. After that there is more time for any new changes.

I've just send you an email.
